File: IPWorks EDI Samples/Secure FTP Client/netcore/ftp.cs
/*
 * IPWorks EDI 2024 .NET Edition - Sample Project
 *
 * This sample project demonstrates the usage of IPWorks EDI in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/ipworksedi
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 * 
 */

﻿using System;
using nsoftware.IPWorksEDI;

class ftpDemo
{
  private static FTP ftp = new nsoftware.IPWorksEDI.FTP();

  private static void ftp_OnSSLServerAuthentication(object sender, FTPSSLServerAuthenticationEventArgs e)
  {
    if (e.Accept) return;
    Console.Write("Server provided the following certificate:\nIssuer: " + e.CertIssuer + "\nSubject: " + e.CertSubject + "\n");
    Console.Write("The following problems have been determined for this certificate: " + e.Status + "\n");
    Console.Write("Would you like to continue anyways? [y/n] ");
    if (Console.Read() == 'y') e.Accept = true;
  }

  private static void ftp_OnTransfer(object sender, FTPTransferEventArgs e)
  {
    Console.WriteLine(e.Text);
  }

  private static void ftp_OnDirList(object sender, FTPDirListEventArgs e)
  {
    Console.WriteLine(e.DirEntry);
  }

  static void Main(string[] args)
  {
    if (args.Length < 3)
    {
      Console.WriteLine("usage: ftp host user pass\n");
      Console.WriteLine("  host  the host to connect to");
      Console.WriteLine("  user  the username to use for authentication");
      Console.WriteLine("  pass  the password to use for authentication");
      Console.WriteLine("\nExample: ftp 192.168.1.2 myusername mypassword");
      Console.WriteLine("Press enter to continue.");
      Console.Read();
    }
    else
    {
      ftp.OnTransfer += ftp_OnTransfer;
      ftp.OnSSLServerAuthentication += ftp_OnSSLServerAuthentication;
      ftp.OnDirList += ftp_OnDirList;

      try
      {
        ftp.RemoteHost = args[0];
        ftp.User = args[1];
        ftp.Password = args[2];

        ftp.Logon();

        Console.WriteLine("Type \"?\" for a list of commands.");
        string command;
        string[] arguments;
        while (true)
        {
          ftp.RemoteFile = "";
          Console.Write("ftp> ");
          command = Console.ReadLine();
          arguments = command.Split();

          if (arguments[0] == "?" || arguments[0] == "help")
          {
            Console.WriteLine("?       cd      ls      pwd");
            Console.WriteLine("get     put     rm      passive");
            Console.WriteLine("mkdir   rmdir   exit");
          }
          else if (arguments[0] == "bye" || arguments[0] == "quit" || arguments[0] == "exit")
          {
            ftp.Logoff();
            break;
          }
          else if (arguments[0] == "cd")
          {
            if (arguments.Length > 1) ftp.ChangeRemotePath(arguments[1]);
          }
          else if (arguments[0] == "get")
          {
            if (arguments.Length > 1)
            {
              ftp.RemoteFile = arguments[1];
              ftp.LocalFile = arguments[1];
              ftp.Download();
              Console.WriteLine("File downloaded");
            }
          }
          else if (arguments[0] == "ls")
          {
            if (arguments.Length > 1)
            {
              string pathname = ftp.QueryRemotePath();
              ftp.ChangeRemotePath(arguments[1]);
              ftp.ListDirectoryLong();
              ftp.ChangeRemotePath(pathname);
            }
            else
            {
              ftp.ListDirectoryLong();
            }
          }
          else if (arguments[0] == "mkdir")
          {
            if (arguments.Length > 1) ftp.MakeDirectory(arguments[1]);
          }
          else if (arguments[0] == "mv")
          {
            if (arguments.Length > 2)
            {
              ftp.RemoteFile = arguments[1];
              ftp.RenameFile(arguments[2]);
            }
          }
          else if (arguments[0] == "passive")
          {
            if (arguments.Length > 1)
            {
              if ((arguments[1] == "on") && !ftp.Passive)
              {
                ftp.Passive = true;
                Console.WriteLine("Passive mode ON.");
              }
              else if ((arguments[1] == "off") && ftp.Passive)
              {
                ftp.Passive = false;
                Console.WriteLine("Passive mode OFF.");
              }
            }
          }
          else if (arguments[0] == "put")
          {
            if (arguments.Length > 2)
            {
              ftp.LocalFile = arguments[1];
              ftp.RemoteFile = arguments[2];
              ftp.Upload();
              Console.WriteLine("File uploaded");
            }
          }
          else if (arguments[0] == "pwd")
          {
            Console.WriteLine(ftp.QueryRemotePath());
          }
          else if (arguments[0] == "rm")
          {
            if (arguments.Length > 1) ftp.DeleteFile(arguments[1]);
          }
          else if (arguments[0] == "rmdir")
          {
            if (arguments.Length > 1) ftp.RemoveDirectory(arguments[1]);
          }
          else if (arguments[0] == "")
          {
            // Do nothing.
          }
          else
          {
            Console.WriteLine("Bad command / Not implemented in demo.");
          } // End of command checking.
        }
      }
      catch (Exception ex)
      {
        Console.WriteLine("Error: " + ex.Message);
      }
      Console.WriteLine("\npress <return> to continue...");
      Console.Read();
    }
  }
}




class ConsoleDemo
{
  /// <summary>
  /// Takes a list of switch arguments or name-value arguments and turns it into a dictionary.
  /// </summary>
  public static System.Collections.Generic.Dictionary<string, string> ParseArgs(string[] args)
  {
    System.Collections.Generic.Dictionary<string, string> dict = new System.Collections.Generic.Dictionary<string, string>();

    for (int i = 0; i < args.Length; i++)
    {
      // Add an key to the dictionary for each argument
      if (args[i].StartsWith("/"))
      {
        // If the next argument does NOT start with a "/" then it is a value.
        if (i + 1 < args.Length && !args[i + 1].StartsWith("/"))
        {
          // Save the value and skip the next entry in the list of arguments.
          dict.Add(args[i].ToLower().TrimStart('/'), args[i + 1]);
          i++;
        }
        else
        {
          // If the next argument starts with a "/", then we assume the current one is a switch.
          dict.Add(args[i].ToLower().TrimStart('/'), "");
        }
      }
      else
      {
        // If the argument does not start with a "/", store the argument based on the index.
        dict.Add(i.ToString(), args[i].ToLower());
      }
    }
    return dict;
  }
  /// <summary>
  /// Asks for user input interactively and returns the string response.
  /// </summary>
  public static string Prompt(string prompt, string defaultVal)
  {
    Console.Write(prompt + (defaultVal.Length > 0 ? " [" + defaultVal + "]": "") + ": ");
    string val = Console.ReadLine();
    if (val.Length == 0) val = defaultVal;
    return val;
  }
}

File: IPWorks EDI Samples/Secure FTP Client/java/ftp.java
/*
 * IPWorks EDI 2024 Java Edition - Sample Project
 *
 * This sample project demonstrates the usage of IPWorks EDI in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/ipworksedi
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 */

import java.io.*;
import ipworksedi.*;

                 

public class ftp
{
    FTP ftp1;
    public int verbose;
    long transtime;
    long transbytes;
    String command;               // user's command
    String[] argument;            // arguments to the user's command
    String pathname;              // for use with the ls command
    public ftp(String[] args){
        try
        {
            ftp1 = new FTP();
            ftp1.addFTPEventListener(new FTPEvents(this));
        }catch (Exception ex)
        {
            System.out.println(ex.getMessage());
        }
        if (args.length >= 3)
        {
            try
            {
                ftp1.setRemoteHost(args[0]);
                ftp1.setUser(args[1]);
                ftp1.setPassword(args[2]);
                ftp1.logon();
            }
            catch(IPWorksEDIException e)
            {
                System.out.println( e.getMessage());
                System.exit(e.getCode());
                return;
            }
        }

        if (args.length == 1)
        {
            try
            {
                ftp1.setRemoteHost(args[0]);
                logon();
            }
            catch(IPWorksEDIException e)
            {
                System.out.println(e.getMessage());
                System.exit(e.getCode());
                return;
            }
        }

        while (true)
        {
            try
            {
                ftp1.setRemoteFile("");
                System.out.print( "ftp> " );
                command = input();
                argument = command.split("\\s");

                if ( argument[0].equals("?") || argument[0].equals("help") )
                {
                    System.out.println("?        bye     help     put     rmdir");
                    System.out.println("append   cd      ls       pwd     verbose");
                    System.out.println("ascii    close   mkdir    quit");
                    System.out.println("binary   get     open     rm");
                }
                else if ( argument[0].equals("append") )
                {
                    ftp1.setLocalFile(argument[1]);
                    ftp1.setRemoteFile(argument[2]);
                    ftp1.append();
                }
                else if ( argument[0].equals("ascii") )
                {
                    ftp1.changeTransferMode(1);
                }
                else if ( argument[0].equals("binary") )
                {
                    ftp1.changeTransferMode(2);
                }
                else if ( argument[0].equals("bye") || argument[0].equals("quit"))
                {
                    ftp1.logoff();
                    System.exit(0);
                    return;
                }
                else if (argument[0].equals("close"))
                {
                    ftp1.logoff();
                }
                else if ( argument[0].equals("cd") )
                {
                    if( argument.length > 0)
                        ftp1.changeRemotePath(argument[1]);
                }
                else if ( argument[0].equals("get") )
                {
                    if ( argument.length < 3 )
                    {
                        System.out.println("get command requires remotefile and localfile.");
                    }
                    else
                    {
                        ftp1.setRemoteFile(argument[1]);
                        ftp1.setLocalFile(argument[2]);
                        ftp1.download();
                        updateTime();
                    }
                }
                else if ( argument[0].equals("ls") )
                {
                    if ( argument.length > 1 )
                    {
                        pathname = ftp1.queryRemotePath();
                        ftp1.changeRemotePath(argument[1]);
                        ftp1.listDirectoryLong();
                        ftp1.changeRemotePath(pathname);
                    }
                    else
                        ftp1.listDirectoryLong();
                }
                else if ( argument[0].equals("mkdir") )
                {
                    if ( argument.length > 1 )
                        ftp1.makeDirectory(argument[1]);
                }
                else if ( argument[0].equals("mv") )
                {
                    ftp1.setRemoteFile(argument[1]);
                    ftp1.renameFile(argument[1]);
                }
                else if ( argument[0].equals("open") )
                {
                    if (argument.length < 2) {
                        System.out.println("open command requires following hostname.");
                    } else {
                        ftp1.logoff();
                        ftp1.setRemoteHost(argument[1]);
                        logon();
                    }
                }
                else if ( argument[0].equals("passive") )
                {
                    if ( argument.length > 1 )
                    {
                        if ((argument[1].equals("on")) && !ftp1.isPassive())
                        {
                            ftp1.setPassive(true);
                            System.out.println( "Passive mode ON." );
                        }
                        else if ((argument[1].equals("off")) && ftp1.isPassive())
                        {
                            ftp1.setPassive(false);
                            System.out.println( "Passive mode OFF." );
                        }
                    }
                }
                else if ( argument[0].equals("put") )
                {
                    if ( argument.length < 3 )
                    {
                        System.out.println("put command requires localfile and remotefile.");
                    }
                    else
                    {
                        // put localfile remotefile
                        ftp1.setRemoteFile(argument[2]);
                        ftp1.setLocalFile(argument[1]);
                        ftp1.upload();
                        updateTime();
                    }
                }
                else if ( argument[0].equals("pwd") )
                {
                    System.out.println( ftp1.queryRemotePath() );
                }
                else if ( argument[0].equals("rm") )
                {
                    if ( argument.length > 1 )
                        ftp1.deleteFile(argument[1]);
                }
                else if ( argument[0].equals("rmdir") )
                {
                    if ( argument.length > 1 )
                        ftp1.removeDirectory(argument[1]);
                }
                else if ( argument[0].equals("verbose") )
                {
                    if ( argument.length > 1 )
                    {
                        if ((argument[1].equals("on")) && verbose == 0)
                        {
                            toggle_verbose();
                        }
                        else if ((argument[1].equals("off")) && verbose == 1)
                        {
                            toggle_verbose();
                        }
                    }
                    else
                    {
                        toggle_verbose();
                    }
                }
                else if ( argument[0].equals("") )
                {
                    // Do nothing
                }
                else {
                    System.out.println( "Bad command / Not implemented in demo." );
                } // end of command checking
            }
            catch(IPWorksEDIException e)
            {
                System.out.println(e.getMessage());
                e.printStackTrace();
                System.exit(e.getCode());
                return;
            }
            catch(IOException e)
            {
                System.out.println(e.getMessage());
            }
        }
    }
    private void updateTime()
    {
        System.out.print(transbytes);
        System.out.print(" bytes sent in ");
        System.out.print(((float) transtime / 1000));
        System.out.print(" seconds.  (");
        System.out.print(((float) transbytes) / transtime);
        System.out.println("KBps)");
    }
    private String input() throws IOException
    {
        BufferedReader bf = new BufferedReader(new InputStreamReader(System.in));
        return bf.readLine();
    }

    void logon() throws IPWorksEDIException
    {
        String buffer;
        try
        {
            System.out.print("User: " );
            buffer = input();
            ftp1.setUser( buffer );
            System.out.print("Password: ");
            buffer = input();
            ftp1.setPassword( buffer );
            ftp1.logon();
        }
        catch(IOException e)
        {
            System.out.println(e.getMessage());
        }
    }
    public void toggle_verbose()
    {
        verbose = 1 - verbose;
        System.out.print( "Verbose mode " );
        if (verbose == 1)
            System.out.println( "ON." );
        else
            System.out.println( "OFF." );
    }

    public static void main(String[] args)
    {
        new ftp(args);
    }
    public void dirList(FTPDirListEvent arg) {
        System.out.println(arg.dirEntry);
    }
    public void PITrail(FTPPITrailEvent arg) {
        if (verbose == 1)
            System.out.println(arg.message);
    }
    public void startTransfer() {
        transtime = System.currentTimeMillis();
    }
    public void transfer(FTPTransferEvent arg) {
        transbytes = arg.bytesTransferred;
    }
    public void endTransfer() {
        long endtime;
        endtime = System.currentTimeMillis();
        transtime = endtime - transtime;
    }
    public void error(FTPErrorEvent arg) {
        System.out.println("\nError "+arg.errorCode+": "+arg.description);
    }
    public void SSLServerAuthentication(FTPSSLServerAuthenticationEvent arg){
        arg.accept = true;
    }
}
class FTPEvents implements FTPEventListener{
		ftp instance;
    public FTPEvents(ftp instance){
        this.instance = instance;
    }
    public void connectionStatus(FTPConnectionStatusEvent arg) {
    }
    public void dirList(FTPDirListEvent arg) {
        instance.dirList(arg);
    }
    public void endTransfer(FTPEndTransferEvent arg) {
        instance.endTransfer();
    }
    public void error(FTPErrorEvent arg) {
        instance.error(arg);
    }
    public void PITrail(FTPPITrailEvent arg) {
    }
    public void startTransfer(FTPStartTransferEvent arg0) {
        instance.startTransfer();
    }
    public void transfer(FTPTransferEvent arg) {
        instance.transfer(arg);
    }
    public void SSLServerAuthentication(FTPSSLServerAuthenticationEvent arg){
        instance.SSLServerAuthentication(arg);
    }
    public void SSLStatus(FTPSSLStatusEvent arg){}
}
class ConsoleDemo {
  private static BufferedReader bf = new BufferedReader(new InputStreamReader(System.in));

  static String input() {
    try {
      return bf.readLine();
    } catch (IOException ioe) {
      return "";
    }
  }
  static char read() {
    return input().charAt(0);
  }

  static String prompt(String label) {
    return prompt(label, ":");
  }
  static String prompt(String label, String punctuation) {
    System.out.print(label + punctuation + " ");
    return input();
  }
  static String prompt(String label, String punctuation, String defaultVal) {
      System.out.print(label + " [" + defaultVal + "] " + punctuation + " ");
      String response = input();
      if (response.equals(""))
        return defaultVal;
      else
        return response;
  }

  static char ask(String label) {
    return ask(label, "?");
  }
  static char ask(String label, String punctuation) {
    return ask(label, punctuation, "(y/n)");
  }
  static char ask(String label, String punctuation, String answers) {
    System.out.print(label + punctuation + " " + answers + " ");
    return Character.toLowerCase(read());
  }

  static void displayError(Exception e) {
    System.out.print("Error");
    if (e instanceof IPWorksEDIException) {
      System.out.print(" (" + ((IPWorksEDIException) e).getCode() + ")");
    }
    System.out.println(": " + e.getMessage());
    e.printStackTrace();
  }
}



